v2/metrics/registry.go

```diff
@@ -0,0 +1,179 @@
+			Name:      "message_handled_total",
+			Help:      "total number of messages handled by this handler",
+			Subsystem: subsystem,
+		}, []string{messageTypeLabel, deliveryCountLabel}),
+		MessageLockRenewedCount: prom.NewCounterVec(prom.CounterOpts{
+			Name:      "message_lock_renewed_total",
+			Help:      "total number of message lock renewal",
+			Subsystem: subsystem,
+		}, []string{messageTypeLabel, successLabel}),
+		MessageDeadlineReachedCount: prom.NewCounterVec(prom.CounterOpts{
+			Name:      "message_deadline_reached_total",
+			Help:      "total number of message lock renewal",
+			Subsystem: subsystem,
+		}, []string{messageTypeLabel}),
+		ConcurrentMessageCount: prom.NewGaugeVec(prom.GaugeOpts{
+			Name:      "concurrent_message_count",
+			Help:      "number of messages being handled concurrently",
+			Subsystem: subsystem,
+		}, []string{messageTypeLabel}),
+	}
+}
+
+func getMessageTypeLabel(msg *azservicebus.ReceivedMessage) prom.Labels {
+	typeName := msg.ApplicationProperties["type"]
+	return map[string]string{
+		messageTypeLabel: fmt.Sprintf("%s", typeName),
+	}
+}
+
+func (m *Registry) Init(reg prom.Registerer) {
+	reg.MustRegister(
+		m.MessageReceivedCount,
+		m.MessageHandledCount,
+		m.MessageLockRenewedCount,
+		m.MessageDeadlineReachedCount,
+		m.ConcurrentMessageCount)
+}
+
+type Registry struct {
+	MessageReceivedCount        *prom.CounterVec
+	MessageHandledCount         *prom.CounterVec
+	MessageLockRenewedCount     *prom.CounterVec
+	MessageDeadlineReachedCount *prom.CounterVec
+	ConcurrentMessageCount      *prom.GaugeVec
+}
+
+// Recorder allows to initialize the metric registry and increase/decrease the registered metrics at runtime.
+type Recorder interface {
+	Init(registerer prom.Registerer)
+	IncMessageDeadlineReachedCount(msg *azservicebus.ReceivedMessage)
+	IncMessageLockRenewedFailure(msg *azservicebus.ReceivedMessage)
+	IncMessageLockRenewedSuccess(msg *azservicebus.ReceivedMessage)
+	DecConcurrentMessageCount(msg *azservicebus.ReceivedMessage)
+	IncMessageHandled(msg *azservicebus.ReceivedMessage)
+	IncMessageReceived(float64)
+	IncConcurrentMessageCount(msg *azservicebus.ReceivedMessage)
+}
+
+// IncMessageLockRenewedSuccess increase the message lock renewal success counter
+func (m *Registry) IncMessageLockRenewedSuccess(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[successLabel] = "true"
+	m.MessageLockRenewedCount.With(labels).Inc()
+}
+
+// IncMessageLockRenewedFailure increase the message lock renewal failure counter
+func (m *Registry) IncMessageLockRenewedFailure(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[successLabel] = "false"
+	m.MessageLockRenewedCount.With(labels).Inc()
+}
+
+// IncMessageHandled increase the message Handled
+func (m *Registry) IncMessageHandled(msg *azservicebus.ReceivedMessage) {
+	labels := getMessageTypeLabel(msg)
+	labels[deliveryCountLabel] = fmt.Sprintf("%d", msg.DeliveryCount)
```

Comment:
super nit: can we use strconv.formatInt here? I think it's more readable (plus a bit faster)